src/librustc_mir/transform/inline.rs

```diff
@@ -0,0 +1,730 @@
+        let mut first_call_in_scc = callsites.len();
+
+        let mut i = 0;
+        while i < first_call_in_scc {
+            let f = callsites[i].caller;
+            if in_scc.contains(&f) {
+                first_call_in_scc -= 1;
+                callsites.swap(i, first_call_in_scc);
+            } else {
+                i += 1;
+            }
+        }
+
+        let mut local_change;
+        let mut changed = false;
+
+        loop {
+            local_change = false;
+            let mut csi = 0;
+            while csi < callsites.len() {
+                let foreign_mir;
+
+                let callsite = callsites[csi];
+                csi += 1;
+
+                let callee_mir = {
+                    let callee_mir : Option<&Mir<'tcx>> = if callsite.callee.is_local() {
+                        map.map.get(&callsite.callee)
+                    } else {
+                        foreign_mir = self.get_foreign_mir(callsite.callee);
+                        foreign_mir.as_ref().map(|m| &**m)
+                    };
+
+                    let callee_mir = if let Some(m) = callee_mir {
+                        m
+                    } else {
+                        continue;
+                    };
+
+                    if !self.should_inline(callsite, callee_mir) {
+                        continue;
+                    }
+
+                    callee_mir.subst(self.tcx, callsite.substs)
+                };
+
+                let caller_mir = map.map.get_mut(&callsite.caller).unwrap();
+
+                let start = caller_mir.basic_blocks().len();
+
+                if !self.inline_call(callsite, caller_mir, callee_mir) {
+                    continue;
+                }
+
+                // Add callsites from inlined function
+                for (bb, bb_data) in caller_mir.basic_blocks().iter_enumerated().skip(start) {
+                    // Only consider direct calls to functions
+                    let terminator = bb_data.terminator();
+                    if let TerminatorKind::Call {
+                        func: Operand::Constant(ref f), .. } = terminator.kind {
+                        if let ty::TyFnDef(callee_def_id, substs, _) = f.ty.sty {
+                            // Don't inline the same function multiple times.
+                            if callsite.callee != callee_def_id {
+                                callsites.push(CallSite {
+                                    caller: callsite.caller,
+                                    callee: callee_def_id,
+                                    substs: substs,
+                                    bb: bb,
+                                    location: terminator.source_info
+                                });
+                            }
+                        }
+                    }
+                }
+
+
+                csi -= 1;
+                if scc.len() == 1 {
+                    callsites.swap_remove(csi);
```

Comment:
@Aatch 

> If scc.len() == 1 then there's only one function in the current SCC, so the order doesn't matter. The order is just between "not in SCC" and "in SCC". The thing is, most SCCs are a single function, so in that case, the order doesn't matter at all. It's just a heuristic, the order doesn't matter for correctness, it just prioritises inlining from outside the SCC first.

OK, that makes sense. This should at minimum be a comment. We can iterate on the overall structure later.